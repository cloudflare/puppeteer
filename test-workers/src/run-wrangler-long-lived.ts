import {fork} from 'node:child_process';
// eslint-disable-next-line no-restricted-imports
import events from 'node:events';
import path from 'node:path';

export const wranglerEntryPath = path.resolve(
  __dirname,
  '../node_modules/.bin/wrangler'
);

/**
 * Runs the command `wrangler dev` in a child process.
 *
 * Returns an object that gives you access to:
 *
 * - `ip` and `port` of the http-server hosting the pages project
 * - `stop()` function that will close down the server.
 */
export async function runWranglerDev(
  cwd: string,
  options: string[],
  env?: NodeJS.ProcessEnv
): Promise<{
  ip: string;
  port: number;
  stop: () => Promise<void>;
  getOutput: () => string;
  clearOutput: () => void;
}> {
  return runLongLivedWrangler(['dev', ...options], cwd, env);
}

async function runLongLivedWrangler(
  command: string[],
  cwd: string,
  env?: NodeJS.ProcessEnv
) {
  let settledReadyPromise = false;
  let resolveReadyPromise: (value: {ip: string; port: number}) => void;
  let rejectReadyPromise: (reason: unknown) => void;

  const ready = new Promise<{ip: string; port: number}>((resolve, reject) => {
    resolveReadyPromise = resolve;
    rejectReadyPromise = reject;
  });
  console.log(command);
  const wranglerProcess = fork(wranglerEntryPath, command, {
    stdio: [
      /* stdin */ 'ignore',
      /* stdout */ 'pipe',
      /* stderr */ 'pipe',
      'ipc',
    ],
    cwd,
    env: {...process.env, ...env, PWD: cwd},
  }).on('message', message => {
    if (settledReadyPromise) {
      return;
    }
    settledReadyPromise = true;
    clearTimeout(timeoutHandle);
    resolveReadyPromise(JSON.parse(message.toString()));
  });

  const chunks: Buffer[] = [];
  wranglerProcess.stdout?.on('data', chunk => {
    chunks.push(chunk);
  });
  wranglerProcess.stderr?.on('data', chunk => {
    chunks.push(chunk);
  });
  const getOutput = () => {
    return Buffer.concat(chunks).toString();
  };
  const clearOutput = () => {
    chunks.length = 0;
  };

  const timeoutHandle = setTimeout(() => {
    if (settledReadyPromise) {
      return;
    }
    settledReadyPromise = true;
    const separator = '='.repeat(80);
    const message = [
      'Timed out starting long-lived Wrangler:',
      separator,
      getOutput(),
      separator,
    ].join('\n');
    rejectReadyPromise(new Error(message));
  }, 50_000);

  async function stop() {
    const closePromise = events.once(wranglerProcess, 'close');
    wranglerProcess.kill('SIGTERM');
    const [code] = await closePromise;
    if (code) {
      throw new Error(`Exited with code ${code}`);
    }
  }

  const {ip, port} = await ready;
  return {ip, port, stop, getOutput, clearOutput};
}
