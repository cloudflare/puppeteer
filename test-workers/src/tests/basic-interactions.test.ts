import {resolve} from 'path';

import {fetch} from 'undici';
import {afterAll, beforeAll, describe, it} from 'vitest';

import {runWranglerDev} from '../run-wrangler-long-lived.js';

describe('browser does basic manipulations', () => {
  let ip: string,
    port: number,
    stop: (() => Promise<unknown>) | undefined,
    getOutput: () => string;

  // @ts-expect-error not sure why this is not working
  beforeAll(async () => {
    return ({ip, port, stop, getOutput} = await runWranglerDev(
      resolve(__dirname, '../..'),
      ['--remote']
    ));
  });

  afterAll(async () => {
    await stop?.();
    console.log(getOutput());
  });

  it('gets the title of google developers website', async ({expect}) => {
    const response = await fetch(`http://${ip}:${port}/basic-interactions`);
    console.log(response);
    expect(await response.text()).toMatchSnapshot();
  });
});
