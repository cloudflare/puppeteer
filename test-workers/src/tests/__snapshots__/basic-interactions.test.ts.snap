// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`browser does basic manipulations > gets the title of google developers website 1`] = `
"<!DOCTYPE html>
<!--[if lt IE 7]> <html class="no-js ie6 oldie" lang="en-US"> <![endif]-->
<!--[if IE 7]>    <html class="no-js ie7 oldie" lang="en-US"> <![endif]-->
<!--[if IE 8]>    <html class="no-js ie8 oldie" lang="en-US"> <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en-US"> <!--<![endif]-->
<head>
<title>Worker threw exception | ai-prices-scrapper.cloudflare-browser-rendering-085.workers.dev | Cloudflare</title>
<meta charset="UTF-8" />
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
<meta http-equiv="X-UA-Compatible" content="IE=Edge" />
<meta name="robots" content="noindex, nofollow" />
<meta name="viewport" content="width=device-width,initial-scale=1" />
<link rel="stylesheet" id="cf_styles-css" href="/cdn-cgi/styles/cf.errors.css" />
<!--[if lt IE 9]><link rel="stylesheet" id='cf_styles-ie-css' href="/cdn-cgi/styles/cf.errors.ie.css" /><![endif]-->
<style>body{margin:0;padding:0}</style>


<!--[if gte IE 10]><!-->
<script>
  if (!navigator.cookieEnabled) {
    window.addEventListener('DOMContentLoaded', function () {
      var cookieEl = document.getElementById('cookie-alert');
      cookieEl.style.display = 'block';
    })
  }
</script>
<!--<![endif]-->


</head>
<body>
  <div id="cf-wrapper">
    <div class="cf-alert cf-alert-error cf-cookie-error" id="cookie-alert" data-translate="enable_cookies">Please enable cookies.</div>
    <div id="cf-error-details" class="cf-error-details-wrapper">
      <div class="cf-wrapper cf-header cf-error-overview">
        <h1>
          <span class="cf-error-type" data-translate="error">Error</span>
          <span class="cf-error-code">1101</span>
          <small class="heading-ray-id">Ray ID: 8abdfc9b3e806918 &bull; 2024-07-31 13:38:49 UTC</small>
        </h1>
        <h2 class="cf-subheadline" data-translate="error_desc">Worker threw exception</h2>
      </div><!-- /.header -->

      <section></section><!-- spacer -->

      <div class="cf-section cf-wrapper">
        <div class="cf-columns two">
          <div class="cf-column">
            <h2 data-translate="what_happened">What happened?</h2>
            <p>You've requested a page on a website (ai-prices-scrapper.cloudflare-browser-rendering-085.workers.dev) that is on the <a href="https://www.cloudflare.com/5xx-error-landing/" target="_blank">Cloudflare</a> network. An unknown error occurred while rendering the page.</p>
          </div>

          
          <div class="cf-column">
            <h2 data-translate="what_can_i_do">What can I do?</h2>
            <p><strong>If you are the owner of this website:</strong><br />you should <a href="https://www.cloudflare.com/login?utm_source=error_100x" target="_blank">login to Cloudflare</a> and check the error logs for ai-prices-scrapper.cloudflare-browser-rendering-085.workers.dev.</p>
          </div>
          
        </div>
      </div><!-- /.section -->

      <div class="cf-error-footer cf-wrapper w-240 lg:w-full py-10 sm:py-4 sm:px-8 mx-auto text-center sm:text-left border-solid border-0 border-t border-gray-300">
  <p class="text-13">
    <span class="cf-footer-item sm:block sm:mb-1">Cloudflare Ray ID: <strong class="font-semibold">8abdfc9b3e806918</strong></span>
    <span class="cf-footer-separator sm:hidden">&bull;</span>
    <span id="cf-footer-item-ip" class="cf-footer-item hidden sm:block sm:mb-1">
      Your IP:
      <button type="button" id="cf-footer-ip-reveal" class="cf-footer-ip-reveal-btn">Click to reveal</button>
      <span class="hidden" id="cf-footer-ip">104.28.246.4</span>
      <span class="cf-footer-separator sm:hidden">&bull;</span>
    </span>
    <span class="cf-footer-item sm:block sm:mb-1"><span>Performance &amp; security by</span> <a rel="noopener noreferrer" href="https://www.cloudflare.com/5xx-error-landing" id="brand_link" target="_blank">Cloudflare</a></span>
    
  </p>
  <script>(function(){function d(){var b=a.getElementById("cf-footer-item-ip"),c=a.getElementById("cf-footer-ip-reveal");b&&"classList"in b&&(b.classList.remove("hidden"),c.addEventListener("click",function(){c.classList.add("hidden");a.getElementById("cf-footer-ip").classList.remove("hidden")}))}var a=document;document.addEventListener&&a.addEventListener("DOMContentLoaded",d)})();</script>
</div><!-- /.error-footer -->


    </div><!-- /#cf-error-details -->
  </div><!-- /#cf-wrapper -->

  <script>
  window._cf_translation = {};
  
  
</script>

</body>
</html>
"
`;
